### ISSUE TITLE
Muon uses Incorrect learning rate for convolutional layers

### ISSUE THREAD

**Issue by hatonosuke** [kozistr/pytorch_optimizer#372]
# Describe the bug

The shape of the parameters is used to determine the learning rate in Muon. 
However, if the parameters are convolutional layers, the shape of the parameters is not in BxA format (B is the number of outputs, A is the number of inputs). 

https://github.com/kozistr/pytorch_optimizer/blob/d4e7564dc58e4fa7bb0f1a1f8d9bfe5f1235116f/pytorch_optimizer/optimizer/muon.py#L205-L207

Therefore, we need to convert the shape of the parameters to BxA format.

In addition, there are modules where the number of inputs and outputs cannot be determined from the shape of parameters alone, such as in_proj_weight in torch.nn.MultiheadAttention and weight in group convolution.

The current implementation does not work as expected for these modules.

**Comment by kozistr** [comment]
@hatonosuke Hi. Thanks for raising this issue! Yes, you're right. The current implementation, which is almost identical to the original implementation, only focuses on the 2d parameter case, especially Linear layer.

To fix that, for 3d and 4d convolutional filters, [we could flatten the last 2/3 dimensions](https://github.com/KellerJordan/Muon/blob/master/muon.py#L48), meaning that `p.size(-2)` stands for the output size of the filter and `p.size(-1)` stands for the rest. I'll work on it soon!

In case of [`in_proj_weight` of `nn.MultiheadAttention` in Pytorch](https://github.com/pytorch/pytorch/blob/v2.6.0/torch/nn/modules/activation.py#L1091), IMHO, I think it could work well likewise the `nn.Linear` layer, cause it has a shape of (output dim, input dim), which is determined at the initialization. Please correct me if I'm wrong.

**Comment by hatonosuke** [comment]
Thank you for fix.

Sorry. I mixed up different problems.

I think that grad of in_proj_weight in MultiheadAttention should be splited to Q,K and V before be passed to zero_power_via_newton_schulz_5.

For example, in the following implementation, the weight is divided into three parts when it is defined.

https://github.com/KellerJordan/modded-nanogpt/blob/64d8eb51ee951e09d9dba09269e0ddc6a099b9a9/train_gpt.py#L262

However, this method requires changes on the model side. It also conflicts with Conv1d.
Therefore, I can't think of a way that works well in all cases.

A similar problem occurs with grouped convolution layers.

**Comment by kozistr** [comment]
> Thank you for fix.
> 
> Sorry. I mixed up different problems.
> 
> I think that grad of in_proj_weight in MultiheadAttention should be splited to Q,K and V before be passed to zero_power_via_newton_schulz_5.
> 
> For example, in the following implementation, the weight is divided into three parts when it is defined.
> 
> https://github.com/KellerJordan/modded-nanogpt/blob/64d8eb51ee951e09d9dba09269e0ddc6a099b9a9/train_gpt.py#L262
> 
> However, this method requires changes on the model side. It also conflicts with Conv1d. Therefore, I can't think of a way that works well in all cases.
> 
> A similar problem occurs with grouped convolution layers.

I got your point and agree that it'd be good to change on the model side, cause it's hard to handle all cases like that.
